src/gpt_j/gpt_j_model.rs

```diff
@@ -0,0 +1,808 @@
+    /// # let vs = nn::VarStore::new(device);
+    /// # let config = GptJConfig::from_file(config_path);
+    /// # let gpt_j_model: GptJModel = GptJModel::new(&vs.root(), &config);
+    /// let (batch_size, sequence_length, past_sequence_length) = (64, 128, 56);
+    /// let input_tensor = Tensor::rand(&[batch_size, sequence_length], (Int64, device));
+    /// let mut past: Vec<Tensor> = Vec::with_capacity(config.n_layer as usize);
+    /// for _ in 0..config.n_layer as usize {
+    ///     past.push(Tensor::rand(
+    ///         &[
+    ///             2,
+    ///             batch_size,
+    ///             config.n_head,
+    ///             past_sequence_length,
+    ///             config.n_embd / config.n_head,
+    ///         ],
+    ///         (Double, device),
+    ///     ))
+    /// }
+    /// let attention_mask = Tensor::zeros(&[batch_size, sequence_length], (Int64, device));
+    /// let token_type_ids = Tensor::ones(&[batch_size, sequence_length], (Int64, device));
+    ///
+    /// let model_output = no_grad(|| {
+    ///     gpt_j_model
+    ///         .forward_t(
+    ///             Some(&input_tensor),
+    ///             Some(&past),
+    ///             Some(&attention_mask),
+    ///             Some(&token_type_ids),
+    ///             None,
+    ///             None,
+    ///             false,
+    ///         )
+    ///         .unwrap()
+    /// });
+    /// ```
+    pub fn forward_t(
+        &self,
+        input_ids: Option<&Tensor>,
+        layer_past: Option<Vec<Option<LayerState>>>,
+        attention_mask: Option<&Tensor>,
+        token_type_ids: Option<&Tensor>,
+        _position_ids: Option<&Tensor>,
+        input_embeds: Option<&Tensor>,
+        train: bool,
+    ) -> Result<GptJModelOutput, RustBertError> {
+        let (calc_input_embeddings, _input_size, _device) =
+            process_ids_embeddings_pair(input_ids, input_embeds, &self.wte)?;
+
+        let input_embeddings =
+            input_embeds.unwrap_or_else(|| calc_input_embeddings.as_ref().unwrap());
+
+        let (layer_past, _layer_past_length) = match layer_past {
+            Some(value) => {
+                assert_eq!(
+                    value.len(),
+                    self.h.len(),
+                    "Past activations vector must be of length equal to the number of layers"
+                );
+                (value.clone(), value.len())
+            }
+            None => {
+                let mut out = Vec::with_capacity(self.h.len());
+                out.resize_with(self.h.len(), || None);
+                (out, 0)
+            }
+        };
+
+        let attention_mask: Option<Tensor> = attention_mask.map(|value| {
+            let attention_mask = value
+                .view((input_embeddings.size()[0], -1))
+                .unsqueeze(1)
+                .unsqueeze(2)
+                .to_kind(input_embeddings.kind());
+
+            let attention_mask: Tensor = (1.0 - attention_mask) * (-10000.0);
```

Comment:
Is this valid for both `fp32` and `fp16`?

Reply:
It looked OK in my tests (same text generated as in Python) with models of both precisions. Do  you see another kind of test for this in particular ?

Reply:
I guess the test you suggested would the `tiny` model will help, I'll let  you know the results.

Reply:
I am not sure if you had the chance to test this - Could you please run the Python version was half precision weights and update the "correctness" integration tests with this the transformation?

Reply:
Well using half precision doesn't work well on CPU (as they usually don't support float16), so this would require Cuda and GPU to run, which is not what the GitHub CI provides I guess.

I will still try to run it on a GPU on my side to check the correctness.

Reply:
I usually try to run a broader set of test using a GPU - maybe a solution could be to add a test for half precision that runs only on GPU? i.e.
```rust
#[test]
fn test_gpt_j_half_precision {
    if Cuda::is_available {
       [... tests here ...]
    }
}
```

Reply:
While giving it a try I realized that with latest Python `transformers` I don't get the same results for the "correctness" test anymore. After investigations it looks like in `4.21.0` HF [made changes](https://github.com/huggingface/transformers/commit/360719a6a4e86c99cb852e844d30712eabb04baf) to the implementation of GPT-J. I will try to adapt the PR accordingly first.

Reply:
Interestingly the generation test is still fine ...